### PR TITLE
GH-29: Implement BorrowItem Lambda Handler

### PR DESCRIPTION
**Pull request by Russell500**

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Default ignored files
```
Comment (yauj) [inline .idea/.gitignore:1]: We don't want to push up the changes for the `.idea` directory. It's IntelliJ config files that aren't useful for other people. I would recommend adding the `.idea` directory in the `gitignore` file, so that it doesn't push it up.

File: amplify/backend/function/BorrowItem/src/package.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "BorrowItem",
+  "version": "2.0.0",
+  "description": "Lambda function generated by Amplify",
+  "main": ",./ts-output/handlers/api/<BorrowItem>.handler",
```
Comment (yauj) [inline amplify/backend/function/BorrowItem/src/package.json:5]: Remove the <> from the string. It should be `./ts-output/handlers/api/BorrowItem.handler`

File: amplify/backend/ts-code/__tests__/api/BorrowItem.test.ts [hunk 3]
```diff
@@ -51,4 +52,35 @@ test('will fail to borrow item when id does not exist', async () => {
         }).then(() => dbClient.getDB())
     ).rejects.toThrow(`Couldn't find item ${TestConstants.BAD_REQUEST} in the database.`)
     expect(dbClient.getDB()).toEqual(DBSeed.ONE_NAME)
+})
+test('will fail to borrow item when item name not passed in', async () => {
```
Comment (yauj) [inline amplify/backend/ts-code/__tests__/api/BorrowItem.test.ts:56]: "will fail to borrow item when item **ids** not passed in"

File: amplify/backend/ts-code/src/api/BorrowItem.ts [hunk 4]
```diff
@@ -46,19 +47,31 @@ export class BorrowItem {
...
         return emitAPIMetrics(
             () => {
-                return Promise.all(scratch.ids.map((id: string) =>
-                    this.mainTable.changeBorrower(id, scratch.borrower, "borrow", scratch.notes)
-                )).then(() => `Successfully borrowed items '${scratch.ids.toString()}'.`)
+                return this.performAllFVAs(input)
+                    .then(()=>
+                        Promise.all(input.ids.map((id: string) =>
+                        this.mainTable.changeBorrower(id, input.borrower, "borrow", input.notes)
+                ))).then(() => `Successfully borrowed items '${input.ids.toString()}'.`)
```
Comment (yauj) [inline amplify/backend/ts-code/src/api/BorrowItem.ts:57]: I would follow the indentation here: https://github.com/gp-seattle/rms-app/blob/master/amplify/backend/ts-code/src/api/ReturnItem.ts#L53-L58. Looks kinda messy.

File: package-lock.json [hunk 5]
```diff
@@ -1,19868 +1,6 @@
 {
```
Comment (yauj) [inline package-lock.json]: Can you revert changes to this file? It should already be setup right. You shouldn't need to override it like this.

**Review by yauj — CHANGES_REQUESTED** [review]
Glad you got it to work! Left some comments.

**Comment by yauj** [comment]
> This pull request **introduces 1 alert** when merging [bfd8d60](https://github.com/gp-seattle/rms-app/commit/bfd8d60097df883da6497d4349af1db2bbdee4c2) into [a1653ec](https://github.com/gp-seattle/rms-app/commit/a1653ec47708c8f84fe365dbd3976e9f69f7dadb) - [view on LGTM.com](https://lgtm.com/projects/g/gp-seattle/rms-app/rev/pr-fa3f35ad4abb2b3b7aeae4a391bc0025a012e173)
> 
> **new alerts:**
> 
> * 1 for Unused variable, import, function or class

@Russell500 Also make sure to address this.

File: .idea/.gitignore (hunk 1 above)
Comment (Russell500) [reply .idea/.gitignore:1]: How would I go about doing this?

File: package-lock.json (hunk 5 above)
Comment (Russell500) [reply package-lock.json]: I can not find this file for my BorrowItem

Comment (Russell500) [reply package-lock.json]: Okay i reverted it to pre commit

File: amplify/backend/function/BorrowItem/src/package.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "BorrowItem",
+  "version": "2.0.0",
+  "description": "Lambda function generated by Amplify",
+  "main": ",./ts-output/handlers/api/BorrowItem.handler",
```
Comment (yauj) [inline amplify/backend/function/BorrowItem/src/package.json:5]: No comma before directory spec. Line should be `"main": "./ts-output/handlers/api/BorrowItem.handler"`

File: amplify/backend/ts-code/__tests__/api/BorrowItem.test.ts [hunk 7]
```diff
@@ -2,7 +2,8 @@ import { BorrowItem } from "../../src/api/BorrowItem"
 import { DBSeed, TestConstants, TestTimestamps } from "../../__dev__/db/DBTestConstants"
 import { LocalDBClient } from "../../__dev__/db/LocalDBClient"
 
-test('will borrow item correctly when id exists', async () => {
+
+test('will borrow item  correctly when id exists', async () => {
```
Comment (yauj) [inline amplify/backend/ts-code/__tests__/api/BorrowItem.test.ts:6]: Can you revert this change?

File: .gitignore [hunk 8]
```diff
@@ -23,4 +23,4 @@ amplifytools.xcconfig
 
 # Backend ts builds
 ts-output/
-report.json
\ No newline at end of file
+report.json'.idea'
```
Comment (yauj) [inline .gitignore:26]: You want to put `.idea` on a different line, like this:

```
ts-output/
report.json
.idea
```

**Review by yauj — CHANGES_REQUESTED** [review]
Can you also remove `package-lock.json` changes?

File: .gitignore [hunk 9]
```diff
@@ -23,4 +24,4 @@ amplifytools.xcconfig
...
```
Comment (yauj) [inline .gitignore:27]: Can you revert this back to just `report.json`?

File: amplify/backend/ts-code/__tests__/api/BorrowItem.test.ts [hunk 10]
```diff
@@ -2,6 +2,7 @@ import { BorrowItem } from "../../src/api/BorrowItem"
 import { DBSeed, TestConstants, TestTimestamps } from "../../__dev__/db/DBTestConstants"
 import { LocalDBClient } from "../../__dev__/db/LocalDBClient"
 
+
```
Comment (yauj) [inline amplify/backend/ts-code/__tests__/api/BorrowItem.test.ts:5]: nit: delete extra line

File: amplify/backend/ts-code/__tests__/api/BorrowItem.test.ts [hunk 11]
```diff
@@ -51,4 +52,35 @@ test('will fail to borrow item when id does not exist', async () => {
         }).then(() => dbClient.getDB())
     ).rejects.toThrow(`Couldn't find item ${TestConstants.BAD_REQUEST} in the database.`)
     expect(dbClient.getDB()).toEqual(DBSeed.ONE_NAME)
+})
```
Comment (yauj) [inline amplify/backend/ts-code/__tests__/api/BorrowItem.test.ts:55]: nit: Add a line between tests.

**Review by yauj — CHANGES_REQUESTED** [review]

File: amplify/backend/backend-config.json [hunk 12]
```diff
@@ -237,6 +237,67 @@
...
+      "providerPlugin": "awscloudformation",
+      "service": "Lambda",
+      "dependsOn": [
+        {
+          "category": "storage",
+          "resourceName": "main",
+          "attributes": [
+            "Name",
+            "Arn",
+            "StreamArn"
```
Comment (yauj) [inline amplify/backend/backend-config.json]: Remove this.

**Review by yauj — APPROVED** [review]
LGTM! Thanks!
